### PR TITLE
Initial implementation

### PR DESCRIPTION
**Pull request by oxtoacart**
For getlantern/lantern-internal#3532

### CODE REVIEW

File: tree.go [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+	return result
+}
+
+func reversed(input string) string {
+	n := 0
+	runes := make([]rune, len(input)+1)
+	// Add a dot prefix to make sure we're only operating on subdomains
+	runes[0] = '.'
+	runes = runes[1:]
+	for _, r := range input {
```
Comment (joesis) [inline tree.go:78]: How about writing runes in reverse order so the final reverse operation can be avoided?

File: tree.go [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+		result[reversed(domain)] = value
+	}
+	return result
+}
+
+func reversed(input string) string {
+	n := 0
+	runes := make([]rune, len(input)+1)
+	// Add a dot prefix to make sure we're only operating on subdomains
+	runes[0] = '.'
```
Comment (joesis) [inline tree.go:76]: I don't quite understand. Can the code reference `runes[0]` by any means after the reassignment below?

**Comment by joesis** [comment]
One more question: Can the domain in `http.Request` be in upper case? If so I guess it needs to be either converted here or by the caller.

File: tree.go (hunk 2 above)
Comment (oxtoacart) [reply tree.go:76]: Oops, I copy and pasted this code, we don't actually need this.

File: tree.go (hunk 1 above)
Comment (oxtoacart) [reply tree.go:78]: I'm just going to totally rewrite this function.

**Comment by oxtoacart** [comment]
> Can the domain in http.Request be in upper case? If so I guess it needs to be either converted here or by the caller.

Yes, good point. I'll make sure I'm doing this in flashlight.

**Comment by oxtoacart** [comment]
@joesis I've added updates per your comments, thanks!

**Comment by oxtoacart** [comment]
Thanks @joesis!

**Pull request merged**